### PR TITLE
Changed character limit to be calculated from character width

### LINKED ISSUE #40463: Block Styles preview name: Character limit is not working correctly for some languages.
### Description

Block Styles preview name is not working properly with character limit depending on language.
Please see the following screenshots.

style

https://github.com/WordPress/gutenberg/tree/trunk/packages/components/src/text#limit

### Step-by-step reproduction instructions

1. Refer to the following code to add a block style.

```PHP

function twenty_twenty_two_register_block_styles() {
	// Hello,Hello,…
	register_block_style(
		'core/image',
		array(
			'name'  => 'twentytwentytwo-image-1',
			'label' => esc_html__( 'Hello,Hello,Hello', 'twentytwentyone' ),
		)
	);

	// Hola,Hola,Ho…
	register_block_style(
		'core/image',
		array(
			'name'  => 'twentytwentytwo-image-2',
			'label' => esc_html__( 'Hola,Hola,Hola', 'twentytwentyone' ),
		)
	);

	// こんにちは,こんにちは
	register_block_style(
		'core/image',
		array(
			'name'  => 'twentytwentytwo-image-3',
			'label' => esc_html__( 'こんにちは,こんにちは', 'twentytwentyone' ),
		)
	);

	// 你好,你好,你好,你好
	register_block_style(
		'core/image',
		array(
			'name'  => 'twentytwentytwo-image-4',
			'label' => esc_html__( '你好,你好,你好,你好', 'twentytwentyone' ),
		)
	);
}
add_action( 'init', 'twenty_twenty_two_register_block_styles' );
```

2. Go to the block editor and place the image block.

3. Open the side panel and look at the style preview.

### Screenshots, screen recording, code snippet

_No response_

### Environment info

- WordPress 5.9.4
- Gutenberg 13.0.0
- Twenty Twenty-Two

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by shimotmk**
Resolves #40463
Related #40807

## What?

Changed character limit to be calculated from character width.

## How?

[This comment](https://github.com/WordPress/gutenberg/issues/40463#issuecomment-1116900354
) is used to calculate the width of the character count using the string-width package.
https://github.com/WordPress/gutenberg/issues/40463#issuecomment-1116900354

### Before
before

### After
image

## Testing Instructions
1. Refer to the following code to add a block style.
```PHP
function twenty_twenty_two_register_block_styles() {
	// Hello,Hello,…
	register_block_style(
		'core/image',
		array(
			'name'  => 'twentytwentytwo-image-1',
			'label' => esc_html__( 'Hello,Hello,Hello', 'twentytwentyone' ),
		)
	);

	// Hola,Hola,Ho…
	register_block_style(
		'core/image',
		array(
			'name'  => 'twentytwentytwo-image-2',
			'label' => esc_html__( 'Hola,Hola,Hola', 'twentytwentyone' ),
		)
	);

	// こんにちは,こんにちは
	register_block_style(
		'core/image',
		array(
			'name'  => 'twentytwentytwo-image-3',
			'label' => esc_html__( 'こんにちは,こんにちは', 'twentytwentyone' ),
		)
	);

	// 你好,你好,你好,你好
	register_block_style(
		'core/image',
		array(
			'name'  => 'twentytwentytwo-image-4',
			'label' => esc_html__( '你好,你好,你好,你好', 'twentytwentyone' ),
		)
	);

	// Hello,こんにちは
	register_block_style(
		'core/image',
		array(
			'name'  => 'twentytwentytwo-image-5',
			'label' => esc_html__( 'Hello,こんにちは', 'twentytwentyone' ),
		)
	);
}
add_action( 'init', 'twenty_twenty_two_register_block_styles' );
```
2. Go to the block editor and place the image block.

3. Open the side panel and look at the style preview.

### CODE REVIEW

File: packages/components/src/truncate/utils.js [hunk 1]
```diff
@@ -2,6 +2,8 @@
  * External dependencies
  */
 import { isNil } from 'lodash';
+// eslint-disable-next-line import/no-extraneous-dependencies
+import stringWidth from 'string-width';
```
Comment (ramonjd) [inline packages/components/src/truncate/utils.js]: Thanks for persisting with this one. 

Things work as described in the browser, and functionality-wise, a good addition to the Truncate component I think.

Screen Shot 2022-06-02 at 2 28 12 pm

My only question would be to ask whether the library is appropriate. 

This package and its dependencies aren't actively maintained, and I'm wondering the utility is worth the bump in package size:

```bash
| `build/components/index.min.js` | 233 kB | +5.01 kB (+2%) |
```

(from the [CI reports](https://github.com/WordPress/gutenberg/runs/6670072210?check_suite_focus=true#step:4:934))

Plus jest doesn't seem to be able to compile the library via Babel.

Even so, do we care about possible ANSI or emoji characters?

Taking those away, it seems as if `string-width` [relies mainly on eastasianwidth](https://github.com/sindresorhus/string-width/blob/main/index.js#L39)

Maybe we could get away with just using https://github.com/komagata/eastasianwidth?

I'm not sure. What do you think?

**Comment by mirka** [comment]
May I suggest we stick with the standard CSS truncation here (the `overflow: hidden` approach [also mentioned](https://github.com/WordPress/gutenberg/issues/40463#issuecomment-1116900354) by @ramonjd)? Using the `Truncate`-based components, we can achieve this by doing:

```jsx
// Assuming `ellipsizeMode='auto'` and `limit=0`, which are the default values
<Truncate numberOfLines={ 1 }>
  Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
</Truncate>
```

My impression is that the character-based `limit` API is not scalable, both in terms of i18n and emojis. Something like `string-width` does compensate for those things, but then... how useful is the `limit` construct really? When would anybody actually want that? I'd think that most single-line truncation needs are based on the (often responsive) rendered width of the container element, not a character-based count/width.

I would probably discourage people from using `limit` unless they absolutely have to, for example if they need `ellipsizeMode` `head` or `middle`. I'm also open to deprecating this prop, as it is easy to misunderstand or misuse, whether or not we add string width compensation. What do folks think?

cc @ciampo

**Comment by ciampo** [comment]
> May I suggest we stick with the standard CSS truncation here (the `overflow: hidden` approach [also mentioned](https://github.com/WordPress/gutenberg/issues/40463#issuecomment-1116900354) by @ramonjd)? Using the `Truncate`-based components, we can achieve this by doing:
> 
> ```js
> // Assuming `ellipsizeMode='auto'` and `limit=0`, which are the default values
> <Truncate numberOfLines={ 1 }>
>   Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
> </Truncate>
> ```
> 
> My impression is that the character-based `limit` API is not scalable, both in terms of i18n and emojis. Something like `string-width` does compensate for those things, but then... how useful is the `limit` construct really? When would anybody actually want that? I'd think that most single-line truncation needs are based on the (often responsive) rendered width of the container element, not a character-based count/width.
> 
> I would probably discourage people from using `limit` unless they absolutely have to, for example if they need `ellipsizeMode` `head` or `middle`. I'm also open to deprecating this prop, as it is easy to misunderstand or misuse, whether or not we add string width compensation. What do folks think?

100% aligned with the above, let's move forward with what suggested by @mirka !

**Comment by shimotmk** [comment]
Hi @mirka
I've made a new pull request based on the code you suggested, so please take a look at it.
#42975

**Pull request closed without merging**